### PR TITLE
Feature/proxies collection

### PR DESCRIPTION
**Pull request by frenchbread**
Closes: #1378 

# Improvements:
* Side menu on `/settings` page. (Using bootstrap's pure layout and styles)
* Unified pages related to settings (e.g. Account, Profile, Branding etc.)
* Project restructure (moved settings related stuff to `/settings` component)
* Changed routing

# Added:
* Proxies collection, schema, permissions

### CODE REVIEW

File: settings/account/client/account.html [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+              </div>
+            </p>
+          </div>
+        </div>
+      </div>
+    </div>
+  </div>
+</template>
+
+<template name="updatePassword">
```
Comment (brylie) [inline settings/account/client/account.html]: Separate this into a different file: `updatePassword.html`

File: settings/account/client/account.html [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    {{> afQuickField name="new" type="password"}}
+    {{> afQuickField name="confirm" type="password"}}
+    <div>
+      <button type="submit" class="btn btn-success">Change Password</button>
+    </div>
+  </fieldset>
+  {{/autoForm}}
+</template>
+
+<template name="setUserName">
```
Comment (brylie) [inline settings/account/client/account.html]: Separate this into a different file: `username.html`

File: settings/profile/client/profile.html [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            {{#autoForm id="updateProfile" collection=Users doc=User type="update"}}
+              {{> afQuickField name="username"}}
+              {{> afQuickField name="profile.company"}}
+              <button type="submit" class="btn btn-success profile-update">
+                {{_ "profile_UpdateButton"}}
+              </button>
+            {{/autoForm}}
+          </div>
+        </div>
+        <div class="panel panel-default">
```
Comment (brylie) [inline settings/profile/client/profile.html]: We will need to hide this section if no proxy has been configured for the system. I.e. a user cannot have an API Key unless one or more proxies has been added.

File: settings/proxies/client/add/add.html [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        <div class="modal-header">
+          <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
+          <h4 class="modal-title" id="addProxyModalLabel">Add Proxy</h4>
+        </div>
+        {{ # autoForm collection=proxiesCollection id="addProxy" type=formType doc=proxy}}
+          <div class="modal-body">
+            {{ > afQuickField name='name' }}
+            {{ > afQuickField name='description' }}
+            {{ > afQuickField name='type' options="allowed" }}
+            {{ # each proxyName in proxiesList }}
```
Comment (brylie) [inline settings/proxies/client/add/add.html]: Interesting loop. We can try this approach in the `ProxyBackends` work.

File: settings/proxies/client/add/add.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  proxiesCollection () {
+    return Proxies;
+  },
+  proxy () {
+    const instance = Template.instance();
+    return instance.data.proxy;
+  },
+  formType () {
+    const instance = Template.instance();
+    return (instance.data.isEdit) ? 'update' : 'insert';
```
Comment (brylie) [inline settings/proxies/client/add/add.js]: Add comment describing the purpose of this line in plain English.

File: settings/proxies/client/proxies.html [hunk 6]
```diff
@@ -0,0 +1,27 @@
+<template name="proxies">
+  <div class="container">
+    <div class="row">
+      <div class="col-md-3 settings-menu">
+        {{ > settingsMenu }}
+      </div>
+      <div class="col-md-6">
+        <h2>
+          <button class="btn btn-default btn-sm pull-right" id="add-proxy">Add proxy</button>
+          Proxies
```
Comment (brylie) [inline settings/proxies/client/proxies.html]: Add i18n tokens.

File: settings/proxies/client/proxies.html [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        </h2>
+        <hr>
+        <div class="">
+          {{ # if proxies.count }}
+            {{ # each proxy in proxies }}
+               {{ > proxyItem proxy=proxy }}
+            {{ / each }}
+          {{ else }}
+            <p class="text-center">
+              <i>No proxies found.</i>
```
Comment (brylie) [inline settings/proxies/client/proxies.html]: Add i18n token.

File: settings/proxies/client/proxies.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+Template.proxies.onCreated(function(){
+  const instance = this;
+
+  instance.proxies = new ReactiveVar();
+
+  instance.subscribe('allProxies');
+
+  instance.autorun(() => {
+    if (instance.subscriptionsReady()) {
+      instance.proxies.set(Proxies.find());
```
Comment (brylie) [inline settings/proxies/client/proxies.js]: Add comment describing the purpose of this line in plain English.

File: settings/proxies/client/proxyItem/proxyItem.html [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+              <th scope="row">API Token</th>
+              <td>*****</td>
+            </tr>
+          </tbody>
+        </table>
+      {{ / if }}
+    </div>
+    <div class="panel-footer">
+      <div class="text-right">
+        <button class="btn btn-success btn-xs" id="edit-proxy">Edit</button>
```
Comment (brylie) [inline settings/proxies/client/proxyItem/proxyItem.html]: Add i18n token.

File: settings/proxies/client/proxyItem/proxyItem.html [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        <button class="btn btn-danger btn-xs" id="remove-proxy">Remove</button>
```
Comment (brylie) [inline settings/proxies/client/proxyItem/proxyItem.html]: Add i18n token.

File: settings/proxies/client/remove/remove.html [hunk 11]
```diff
@@ -0,0 +1,19 @@
+<template name="removeProxy">
+  <div class="modal fade" id="removeProxyModal" tabindex="-1" role="dialog" aria-labelledby="removeProxyModalLabel">
+    <div class="modal-dialog" role="document">
+      <div class="modal-content">
+        <div class="modal-header">
+          <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
+          <h4 class="modal-title" id="removeProxyModalLabel">Are you sure that you want to remove proxy?</h4>
```
Comment (brylie) [inline settings/proxies/client/remove/remove.html]: Add i18n token.

File: settings/proxies/client/remove/remove.html [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+        </div>
+        <div class="modal-body">
+          {{ proxy.name }}
+        </div>
+        <div class="modal-footer">
+          <button type="submit" class="btn btn-danger" id="confirm-remove-proxy">Yes, remove</button>
```
Comment (brylie) [inline settings/proxies/client/remove/remove.html]: Add i18n token.

File: settings/proxies/client/remove/remove.html [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+          <button type="button" class="btn btn-default" data-dismiss="modal">Cancel</button>
```
Comment (brylie) [inline settings/proxies/client/remove/remove.html]: Add i18n token.

File: settings/proxies/collection/permission.js [hunk 14]
```diff
@@ -0,0 +1,14 @@
+import { Proxies } from './';
+
+Proxies.allow({
+  insert: function () {
+
+    return Roles.userIsInRole(Meteor.userId(), ['admin']);
```
Comment (brylie) [inline settings/proxies/collection/permission.js]: Add plain English comment describing the purpose of this line. That way we don't have to parse JavaScript in our brain to figure out the goal here.

File: settings/proxies/collection/permission.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+  },
+  update: function () {
+    return Roles.userIsInRole(Meteor.userId(), ['admin']);
```
Comment (brylie) [inline settings/proxies/collection/permission.js]: Add plain English comment describing the purpose of this line. That way we don't have to parse JavaScript in our brain to figure out the goal here.

File: settings/proxies/collection/permission.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+  },
+  remove: function () {
+    return Roles.userIsInRole(Meteor.userId(), ['admin']);
```
Comment (brylie) [inline settings/proxies/collection/permission.js]: Add plain English comment describing the purpose of this line. That way we don't have to parse JavaScript in our brain to figure out the goal here.

File: settings/proxies/default/schemas/apiUmbrella.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
+export default {
+  apiUmbrella: {
+    type: Object,
+    optional: true
```
Comment (brylie) [inline settings/proxies/default/schemas/apiUmbrella.js]: Add custom validation function to require this object/sub fields if the 'type' field equals 'apiUmbrella'.

File: settings/proxies/default/schemas/kong.js [hunk 18]
```diff
@@ -0,0 +1,18 @@
+export default {
+  kong: {
```
Comment (brylie) [inline settings/proxies/default/schemas/kong.js]: Add custom validation function to require this object/sub-fields if the 'type' field equals 'kong'.

File: settings/proxies/default/schemas/tyk.js [hunk 19]
```diff
@@ -0,0 +1,18 @@
+export default {
+  tyk: {
```
Comment (brylie) [inline settings/proxies/default/schemas/tyk.js]: Add custom validation function to require this object/sub-fields if the 'type' field equals 'tyk'.

File: settings/proxies/server/publications.js [hunk 20]
```diff
@@ -0,0 +1,5 @@
+import { Proxies } from '../collection';
+
+Meteor.publish('allProxies', () => {
+  return Proxies.find();
```
Comment (brylie) [inline settings/proxies/server/publications.js]: We may want to check that a user has admin role within this publication. Return an empty array if the user is not an admin.

File: settings/settings/client/menu/menu.html [hunk 21]
```diff
@@ -0,0 +1,11 @@
+<template name="settingsMenu">
+  <ul class="list-group">
+    <a href="{{ pathFor route='settings.profile' }}" class="list-group-item {{ isActive 'settings.profile' }}">Profile<span class="glyphicon glyphicon-chevron-right pull-right"></span></a>
```
Comment (brylie) [inline settings/settings/client/menu/menu.html]: Add i18n token.

File: settings/settings/client/menu/menu.html [hunk 22]
```diff
@@ -0,0 +1,11 @@
...
+    <a href="{{ pathFor route='settings.account' }}" class="list-group-item {{ isActive 'settings.account' }}">Account<span class="glyphicon glyphicon-chevron-right pull-right"></span></a>
```
Comment (brylie) [inline settings/settings/client/menu/menu.html]: Add i18n token.

File: settings/settings/client/menu/menu.html [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+    {{ # if isInRole 'admin' }}
+      <a href="{{ pathFor route='settings.branding' }}" class="list-group-item {{ isActive 'settings.branding' }}">Branding<span class="glyphicon glyphicon-chevron-right pull-right"></span></a>
```
Comment (brylie) [inline settings/settings/client/menu/menu.html]: Add i18n token.

File: settings/settings/client/menu/menu.html [hunk 24]
```diff
@@ -0,0 +1,11 @@
...
+      <a href="{{ pathFor route='settings' }}" class="list-group-item {{ isActive 'settings' }}">Settings<span class="glyphicon glyphicon-chevron-right pull-right"></span></a>
```
Comment (brylie) [inline settings/settings/client/menu/menu.html]: Add i18n token.

File: settings/settings/client/menu/menu.html [hunk 25]
```diff
@@ -0,0 +1,11 @@
...
+      <a href="{{ pathFor route='settings.proxies' }}" class="list-group-item {{ isActive 'settings.proxies' }}">Proxies<span class="glyphicon glyphicon-chevron-right pull-right"></span></a>
```
Comment (brylie) [inline settings/settings/client/menu/menu.html]: Add i18n token.

File: settings/settings/client/menu/menu.js [hunk 26]
```diff
@@ -0,0 +1,7 @@
+import { Template } from 'meteor/templating';
+
+Template.settingsMenu.helpers({
+  isActive (route) {
```
Comment (brylie) [inline settings/settings/client/menu/menu.js]: This is redundant, since we already have the `zimme:iron-router-active` package in our app.

https://github.com/zimme/meteor-active-route#isactiveroute

**Comment by brylie** [comment]
Overall, the structure of this PR looks good. There are a few notes:
- check security inside publication
- remember i18n tokens
- always add plain language comments before any significant line of code

**Comment by brylie** [comment]
We also need custom validation functions in the schema, so that fields are required based on the proxy type.

**Comment by brylie** [comment]
Consider also what happens if no proxy has been configured for the system. E.g. the API key section of the user profile is not relevant in that scenario.

**Comment by brylie** [comment]
Remember also to move *any* API Umbrella or settings/account related code to a sub-component (directory) of the proxies directory. This includes:

- /server/methods/apiUmbrella.js
- /server/publish/apiUmbrella*
- /server/publish/user.js
- server/accounts.js

Basically, we want to start moving everything out of the `/server/` directory and into component specific directories.

Also, feel free to remove any API Umbrella, settings, or accounts code that you feel is not useful.

**Comment by frenchbread** [comment]
@brylie Please help with `custom validation functions in the schema`.

**Comment by brylie** [comment]
Cool. When will you be back in the office?

**Comment by brylie** [comment]
Remember also to remove any API Umbrella / Elasticsearch settings from the settings collection. Also, clean up the Settings section of the Settings page. Rename the `Settings` section to be more specific, such as only Mail settings, for instance.

**Comment by brylie** [comment]
Note that the "Initial Setup Wizard" is being deprecated in my pull request. This wizard is not necessary any longer, since we are not depending so heavily on API Umbrella being available.

**Comment by Nazarah** [comment]
@frenchbread : can you add a * mark beside the field names s.t. users at first glance can understand they need the fill them up instead waiting for alert text to appear and notify them about this?

File: settings/proxies/default/schemas/apiUmbrella.js [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  'apiUmbrella.url': {
```
Comment (brylie) [inline settings/proxies/default/schemas/apiUmbrella.js]: Add RegEx URL validation to this field.

https://github.com/aldeed/meteor-simple-schema#regex

File: settings/proxies/default/schemas/kong.js [hunk 28]
```diff
@@ -0,0 +1,18 @@
...
+    type: Object,
+    optional: true
+  },
+  'kong.url': {
```
Comment (brylie) [inline settings/proxies/default/schemas/kong.js]: Add RegEx URL validation to this field.

https://github.com/aldeed/meteor-simple-schema#regex

File: settings/proxies/default/schemas/tyk.js [hunk 29]
```diff
@@ -0,0 +1,18 @@
...
+    type: Object,
+    optional: true
+  },
+  'tyk.url': {
```
Comment (brylie) [inline settings/proxies/default/schemas/tyk.js]: Add RegEx URL validation to this field.

https://github.com/aldeed/meteor-simple-schema#regex

**Comment by frenchbread** [comment]
@Nazarah Ok

@brylie I'll be in the office next sprint.

**Comment by frenchbread** [comment]
@Nazarah We need to decide what fields are actually optional because I see only `description` field as optional since others are required for proxy to work. API Umbrella's `elasticsearch` field also could be optional. @brylie What do you think?

File: settings/profile/client/profile.html (hunk 3 above)
Comment (frenchbread) [inline settings/profile/client/profile.html]: Should we remove this block totally from here? What is the use case when it's needed? In updated version of schemas API key could be fetched directly from proxies collection.

**Comment by brylie** [comment]
Description could be optional. All API Umbrella fields should be required if the proxy `type === 'apiUmbrella'`. We need a custom validator as well as user interface elements to inform the user that fields are required.

**Comment by jykae** [comment]
@frenchbread @brylie I made PR where I can integrate our work #1432

**Comment by jykae** [comment]
great work @frenchbread this integrated without conflicts! 👍 

I've been thinkings that as Analytics dashboard is Umbrella feature, we could possibly enable/disable it based on if Umbrella proxy is available? Or should it be possible for admin to also toggle that?

File: settings/proxies/default/schemas/tyk.js (hunk 29 above)
Comment (jykae) [inline settings/proxies/default/schemas/tyk.js]: @brylie @frenchbread Do we know that Tyk uses same kind of auth? Does it have Admin API?

**Comment by frenchbread** [comment]
@jykae Good idea, I think we can do that now or when we will add "real" support for other proxies than APIUmbrella. Ideally, (since other proxies may also provide analytics) we could rewrite our dashboard to fit all possible scenarios for analytics data buckets.

**Comment by frenchbread** [comment]
@brylie I've tried to implement custom validation following the same principle as on dipor but noticed that there are some parts of code that require additional error handling etc. Could you help  with that? I don't think it'll take much time and also implementing validation for one field and copy-pasting it for others should work fine. :grin:

**Comment by brylie** [comment]
Jep. When would you like to pair program? Will you be in the office this week?

**Comment by frenchbread** [comment]
@brylie I meant if you could do it yourself, because I don't have Apinf time left for this sprint. Otherwise I can add custom validation next week myself.

File: settings/proxies/default/schemas/tyk.js (hunk 29 above)
Comment (frenchbread) [inline settings/proxies/default/schemas/tyk.js]: @jykae Not sure about it. It's just a sample schema for now.

**Comment by brylie** [comment]
As per our chat, we will not focus on custom validation for the time being.

**Comment by bajiat** [comment]
@frenchbread: @jykae will fix the form

**Comment by jykae** [comment]
@brylie Is the form selection (type Umbrella) only thing needed to fix? Or did you discuss something else also with @frenchbread ?

**Comment by brylie** [comment]
The form is the only necessary fix. Basically, the form should be static. Behind the scenes, the type field will default to 'apiUmbrella' for the time being.

**Comment by jykae** [comment]
@brylie Thanks for pair of eyes 👍

**Comment by jykae** [comment]
This combining schemas solution is kinda interesting 😕 
@brylie Would removing options from the list be ok? I am trying to think how to require Umbrella fields when that type is selected. Or was that validating decided to left out for now?

**Comment by jykae** [comment]
@brylie ok, ready for review

**Comment by frenchbread** [comment]
I think if we want to use only one schema for now (e.g. apiUmbrella's) then we would need remove or comment code that takes care of managing schemas dynamically to have only one schema in `/settings/proxies/collection/schema.js`

:point_up: Removing everything but `apiUmbrella` from `list.js` would also do the trick but user'll have to manually select apiUmbrella from list anyways.

**Comment by frenchbread** [comment]
Checked commits, and yeah, pretty much what I meant. Thank for finishing this one @jykae! 👍

**Comment by brylie** [comment]
Looks good. Merging, so I can keep working on the project restructuring with these changes.
